### ISSUE TITLE
PM Tickets Random Numbers Lifecycle

### ISSUE THREAD

**Issue by eladmallel** [livepeer/go-livepeer#620]
**Is your feature request related to a problem? Please describe.**

__The Basics:__

* PM is based on B sending O payment tickets that are only redeemable under a certain probability (the probability is subject to B & O's negotiation).
* The function that determines if a ticket is redeemable (i.e. a winning ticket) takes two random numbers as inputs: one from B and one from O, named `senderNonce` and `recipientRand` respectively.

__Problem/Solution Flow:__

**Problem:** An obvious attack vector to defend against is B sending `senderNonce`s such that no ticket is redeemable. 
**Solution:** provide B with O's commitment to a `recipientRand` without revealing it to B, thus only O can determine for any `senderNonce` whether a given ticket is redeemable.

**Problem:** A new problem arises when O submits winning tickets to the `TicketBroker` smart contract: O must reveal `recipientRand` for the contract to verify the ticket is a winner before it settles the payment, thus making `recipientRand` visible to B as well. 
**Solution:** O must generate a new `recipientRand` every time O reveals it on chain, and ask all Bs to use this new `recipientRand` hence forth.

**Problem:** The next problem arises from the asynchronous nature of B's and O's operations, e.g. B might be in the middle of sending a ticket with the old `recipientRand` commitment just as O is broadcasting the new `recipientRand`. 
**Solution:** We think O should provide some grace, allowing B to use the old value for some `N` number of tickets before, to provide a continuous transcoding service for all honest Bs.

**Problem:** The next problem is that given say a grace of `N=3` invalid tickets in a row, B could exploit O by sending an invalid ticket every other ticket, which would conform to the `N=3` threshold, but would result in 50% of tickets being invalid, which is obviously unwelcome. 
**Solution:** O should maintain a global ratio of `# invalid tickets / # all tickets` and a configurable threshold for this ratio. By default O will block any B that crosses this threshold.

**Problem:** If B knows O's threshold, B could game O by sending invalid tickets just below O's threshold. 
**Solution:** log (either to DB or a log file) all Bs' invalid ticket ratios to allow O's admins to monitor for such behaviors and potentially decide to block specific Bs. **TODO**: like to the relevant issue here

**Problem:** Finally B could attempt to reuse `senderNonce` values it believes did not yield a winning ticket for O (for the same `recipientRand` commitment). 
**Solution:** O will protect against this attack by maintaining a list of all `senderNonce` values each B used for the current `recipientRand` value. Once O generates a new `recipientRand` value and broadcasts it, O can reset its `senderNonce` state.

Future concern:
* If and when we decouple tickets from segments (i.e. allowing a ticket to pay for multiple segments), we will also have to protect against B not sending enough tickets on time (e.g. with some timeout).

Open questions:
* Should O publish the invalid tickets ratio threshold in `OrchestratorInfo`?
- Should O have some mechanism to allow blocked Bs to redeem themselves and do business with O after being blocked?

**Describe the solution you'd like**

* O generates a new `recipientRand` every time is boots.
* Any B wishing to work with O receives O's current `recipientRand` value as part of the `OrchestratorInfo` message.
* When O decides to submit a winning ticket on-chain, O generates a new `recipientRand` value and broadcasts a new `OrchestratorInfo` message with the new value to all Bs currently streaming to O.
* O maintains a list of used `senderNonce`s for each streaming B, and resets this list upon generating a new `recipientRand` value.
* O allows B to send `invalidTicketsInARowAllowed` invalid tickets in a row before disconnecting, to allow B time to receive and use the new `recipientRand` value (default setting `invalidTicketsInARowAllowed=3`).
* O maintains (on disk) an ongoing invalid tickets ratio per B as well as a `invalidTicketRatioThreshold` above which O rejects further video/tickets from B (`invalidTicketRatioThreshold` should be at least `invalidTicketsInARowAllowed/ticketWinningProbability`)

**Describe alternatives you've considered**

* We considered enforcing ticket order, but since currently tickets and segments are coupled, enforcing such order would hurt video performance, and since video performance is top priority, this solution seemed unfavorable.
* We considered the HMAC approach to hold less state in O, but concluded that HMAC doesn't provide protection against Bs replaying old `recipientRand` commitments to continuously generate losing tickets. To prevent this attack we concluded that O must reject previous `recipientRand` values.

**Comment by yondonfu** [comment]
> Should O publish the invalid tickets ratio threshold in OrchestratorInfo?

I don't think this is necessary as this information should be internal to O.

> Should O have some mechanism to allow blocked Bs to redeem themselves and do business with O after being blocked?

For the sake of minimizing complexity, I would say no. A B can always use a different ETH address if it needs to (assuming the block list is based on ETH addresses).

> O maintains (on disk) an ongoing invalid tickets ratio per B as well as a invalidTicketRatioThreshold above which O rejects further video/tickets from B (invalidTicketRatioThreshold should be at least invalidTicketsInARowAllowed/ticketWinningProbability)

We can leave something like this for SS and use the simplest approach of accepting `invalidTicketsInARowAllowed = 3` invalid tickets before disconnecting for MVS, yeah?

> We considered the HMAC approach to hold less state in O, but concluded that HMAC doesn't provide protection against Bs replaying old recipientRand commitments to continuously generate losing tickets. To prevent this attack we concluded that O must reject previous recipientRand values.

Adding a little more background to this decision:

There are 2 approaches to defending against sender replay attacks that use tickets with `recipientRandHash` values with already revealed preimages: 1) O keeps track of its current `recipientRand` value and considers any tickets received with `recipientRandHash != H(recipientRand)` as invalid 2) O does not keep track of its current `recipientRand` value and instead keeps track of a set of old (invalid) `recipientRand` values. 

The goal of the proposed HMAC construction is statelessness: O maintains a local `randSecret` and sends `randSeed` to Bs s.t. O can always derive `recipientRand` as `HMAC(randSecret, randSeed)`. Approach 1 would make the HMAC construction unnecessary - O would keep track of its current `randSeed` value (to derive its current `recipientRand` value). In this case, O might as well just skip the HMAC construction altogether and just keep track of its current `recipientRand` value. Approach 2 allows O to avoid keeping track of its current `randSeed` value, but actually requires O to keep track of even more state in the form of the set of old (invalid) `recipientRand` values.

Approach 1 seems preferable since it requires tracking the least amount of state - a single in-memory value for the current `recipientRand` and if we're going to use approach 1 then we can skip the HMAC construction.

**Comment by j0sh** [comment]
> The next problem is that given say a grace of N=3 invalid tickets in a row

The grace could be an absolute count, rather than "in a row".

>  O must generate a new recipientRand every time O reveals it on chain, and ask all Bs to use this new recipientRand hence forth.

Sharing a single `recipientRand` for all Bs has quite a few spillover effects in addition to the ticket-validity issues outlined. We should consider using multiple `recipientRand`s.

* The broadcaster can keep its ticket-handling isolated on a per-stream basis with multiple `recipientRand`.  Eg, a B node sending multiple streams into an O would have separate values for `recipientRand` for each stream. Resetting any one `recipientRand` avoids the need for B to reset its ticket(s) for all streams using that O. This mitigates a requirement for global PM synchronization on B, since we can't guarantee that it has fewer than N streams using O.

* Since there are uncertain interactions on the broadcaster side, we can be far stricter about the "grace period" for invalid tickets. B should immediately send tickets with updated parameters, modulo the segments that are already in-flight. This predictability should also increase reliability.

* In general, multiple `recipientRand` allows B to be better-shielded from the details of O's PM construction and reduces the overall complexity of a watertight implementation.

*  With a single `recipientRand`, there needs to be a global lock around the orchestrator's PM handler, making it a bottleneck in the overall flow that's shared by all streams.

* With a single `recipientRand`, Bs that generate winning tickets more frequently means tickets are invalidated more frequently for *all* other Bs using the same orchestrator, which may not be desirable.

A stateless HMAC effectively gives us an unbounded number of `recipientRand` values while minimizing the incidental complexity mentioned above.

* Each broadcaster effectively gets a new `recipientRand` each time it requests pricing information: the orchestrator simply generates a new `randSeed` and derives `recipientRand` based on that.

* Since pricing information will often be fetched speculatively; we don't need to allocate and track state for `recipientRand` for each pricing request.

* This does have the drawback that we might need to store more `senderNonce` values, as not all streams with a given `recipientRand` will result in a winning ticket. However, this isn't necessarily "additional" bookeeping that we have to implement; it's the same bookkeeping as single-`recipientRand` (indexed differently), but with some data likely to be retained for a longer period of time.

* The above drawback is also a potential attack vector, but that can be mitigated later at a higher layer. Eg, scanning logs and blocking a B that seems to be requesting and/or using too many fresh `recipientRand` values.

> Approach 2 allows O to avoid keeping track of its current randSeed value, but actually requires O to keep track of even more state in the form of the set of old (invalid) recipientRand values.

Only for as long as `recipientRand` can be derived from the current `randSecret`. So if `randSecret` is reset (such as a node restart) then the old values don't have to be tracked anymore. This is also only necessary for winning tickets, which will be relatively few and far between.

Again, this presumes that we can have multiple `recipientRand`s. Of course, if you still opt to maintain a single `recipientRand` for all broadcasters, then Approach 1 makes more sense. But I don't think it's really good for the long run.

**Comment by eladmallel** [comment]
@j0sh can you outline the entire HMAC solution? - I'm pretty sure I'm not really getting how you're thinking about it :)

**Comment by eladmallel** [comment]
> Sharing a single recipientRand for all Bs has quite a few spillover effects in addition to the ticket-validity issues outlined. We should consider using multiple `recipientRand`s

I connect mostly with your first and last points, how a rand-per-stream minimizes effects on other streams, whether from the same B or from other Bs.

I think it's a good idea to minimize the frequency of `recipientRand` changes across the board, because every such switch costs O some amount of invalid tickets that might be sent with the old value.

Therefore I am convinced of the idea of a `recipientRand` per stream.

I can't really reason around the HMAC solution until I feel confident that we're talking about the same thing - so waiting on your reply (Josh) to address that idea fully.

Thanks for the feedback @j0sh!

**Comment by j0sh** [comment]
@eladmallel  Absolutely.

Each time O starts up, it generates a fresh `randSecret`. (The `randSecret` should not be persisted across restarts due to the tracking required for winning tickets.)

Each time B requests PM parameters, O generates a fresh `randSeed` and derives `recipientRand` from that, eg `recipientRand = HMAC(randSecret, randSeed)`

The primary motivation for the `recipientRand` construction using HMAC is that PM parameters may often be fetched speculatively, so we don't want to retain any long-lived state for each request. Hence, HMAC allows us to have a per-broadcaster `recipientRand` while validating the PM ticket at some later time with minimal state tracking -- only the `randSecret` is required.

Note that we can incorporate additional parameters in the HMAC construction, eg the `winProb` or `faceValue`. Maybe we want to validate all the parameters that O sends (actually, we probably should), or maybe these parameters are expensive to compute (eg, requires a remote lookup) and we want the ticket processing to be as quick as possible. Here is an example:

```
func GenPMParams(broadcasterAddr) PMParams {
  randSeed = random()
  winProb = getWinProbFor(broadcasterAddr)
  faceValue = getFaceValueFor(broadcasterAddr)
  recipientRand = HMAC(randSecret, randSeed|winProb|faceValue)
  params = PMParams {
    recipientRandHash = H(recipientRand),
    randSeed = randSeed,
    winProb = winProb,
    faceValue = faceValue,
    ... other fields ...
  }
  return params
}
```

Ticket verification can then be done mostly statelessly with respect to prior knowledge B. If `recipientRand` passes muster, then we can be assured that the parameters originate from us, and weren't modified by anyone.

```
func CheckTicket(ticket) recipientRand, error {
  recipientRand = HMAC(randSecret, ticket.randSeed|ticket.winProb|ticket.faceValue) 
  if H(recipientRand) != ticket.recipientRandHash return nil, InvalidTicket
  if PreviousWinner(recipientRand) return nil, InvalidTicket
  if NonceReused(recipientRand, ticket.senderNonce) return nil, InvalidTicket
  ... other checks ...
  return recipientRand, nil
}
func PreviousWinner(recipientRand) {
  // Use a threadsafe implementation of a set
  return SetContains(winningTickets, recipientRand)
}
func NonceReused(recipientRand, senderNonce) {
  return SetContains(senderNonces[recipientRand], senderNonce)
}
```

Then the ticket can be processed, like so:

```
func ProcessTicket(ticket) {
  recipientRand, err = CheckTicket(ticket)
  if err != nil return err
  AddToSet(senderNonces[recipientRand], ticket.senderNonce)
  if TicketWins(ticket) {
    AddToSet(winningTickets, recipientRand)
    ClearSet(senderNonces[recipientRand])
  }
    ... other processing ...
}
```

Note that the only time retain state is when we actually receive a ticket. Since we're likely to get paid, it's okay to incur this storage cost.

Winning tickets also need to be tracked for as long as `randSecret` is active. Should `randSecret` be reset, all `recipientRands` would immediately be invalidated.

As noted, this approach also isn't without its drawbacks, but I feel it avoids much of the incidental complexity that would be required to ensure a watertight implementation with a single `recipientRand`, such as the spillover with B's handling of invalidated tickets.

**Comment by j0sh** [comment]
> When O decides to submit a winning ticket on-chain, O generates a new recipientRand value and broadcasts a new OrchestratorInfo message with the new value to all Bs currently streaming to O.

The notion of O "broadcasting" to B is a bit tricky here, since we don't actually have bidirectional communications -- B -> O follows a client-server, request-reply format due to limitations in gRPC. While there are workarounds for this, there is perhaps a simpler way.

If O detects an invalid ticket within a segment, it can return updated `OrchestratorInfo` parameters  along with the transcoded results for the segment. We can put the `OrchestratorInfo` in either the HTTP headers (since this segment submission is done over plain HTTP2), or the [`TranscodeResult`](https://github.com/livepeer/go-livepeer/blob/a8362ac3655062328d95be07976d5f90ccbc136f/net/lp_rpc.proto#L130) protobuf struct, which gets sent down in response to the segment submission. I'm slightly in favor of the latter.

This would also work nicely as a "redirect" mechanism, since the `OrchestratorInfo` data contains the URI of the orchestrator that is supposed to be used for the following segments. That would allow orchestrators to redirect load among one another.

If a B gets redirected, this would imply yet another round-trip to receive updated PM parameters from the "new" orchestrator node (multiple nodes belonging to the same orchestrator operator probably still want separate PM / `randSecret` params to maintain bookkeeping independence), but we'd get that behavior for free with the segment submission flow. Feels like it'd work nicely.

**Comment by yondonfu** [comment]
Here is a summary of an offline discussion between @eladmallel and I:

We agree that it makes more sense to have different `recipientRand` values instead of a single global `recipientRand` value in order to reduce the complexity issues mentioned previously. Thus, we compare the approach suggested in the OP of this issue modified to use different `recipientRand` values per stream (instead of using a single global `recipientRand` value) with the HMAC construction.

**Approach suggested in OP of this issue**

Since Bs can speculatively submit pricing requests, O must persist a `recipientRand` value even O is not actively working with B yet. This increases memory requirements for O. One way to mitigate this is for O to periodically clean up memory that is not being used to store `recipientRand` values for Bs that O is actively working with.

However, related to the above point, O might still be vulnerable to a memory consumption attack. Suppose O cleans up memory every 10 minutes. Within this 10 minute time window, an attacker can flood O with speculative pricing requests that force O to store many `recipientRand` values. O can defend against this type of memory consumption attack with techniques such as IP and ETH address throttling.

**HMAC Construction**

No additional state tracking when responding to speculative price requests.

The set of old `recipientRand` values for winning tickets maintained by O must be global i.e. O cannot simply maintain a set of old `recipientRand` values for winning tickets per stream because B1 could use a old `recipientRand` value for a winning ticket sent by B2 which would allow B1 to generate non-winning tickets that O does not detect since this particular `recipientRand` value is not included in the `recipientRand` set for B1's stream. There is a requirement of ensuring threadsafe access of this global `recipientRand` set.

Furthermore, for Os that stay online for longer periods of time (seems to be fair to assume that this will be a common case), the `recipientRand` set can grow fairly large. O might choose to refresh `randSecret` periodically in order to reset the `recipientRand` set.

The cons for the approach from the OP of this issue seems to incur greater complexity overhead than the HMAC construction due to the possibility of the memory consumption attack on O. Thus, we are currently leaning toward using the HMAC construction.

**Comment by j0sh** [comment]
> O might still be vulnerable to a memory consumption attack. 

To be fair, the HMAC construction is also vulnerable to a similar attack. In particular, a B can obtain many PM parameters, and send a steady stream of segments, each with a different set of PM parameters. O would have no choice but to store all these [1]. The one saving grace is that B would eventually be paying O for this trouble.

The memory consumption attack for both approaches can be mitigated by higher-level monitoring, and constant-space set implementations such as a Bloom filter. The PM parameters themselves could also have an expiry, which should be incorporated in the construction of `recipientRand` (described below).

O also has a few other approaches at its disposal to address the issue at segment-submission time. One is to only allow a reasonable number of concurrent streams/"jobs" from each B (say, 100). Since each job (say jobs are indexed by `recipientRand`) still needs a certain amount of state tracking for accounting, and this state tracking has a finite lifetime (we can just use the existing transcode loop for this: 10 minutes), the concurrency limit effectively throttles the growth of the in-memory state space associated with a given B (100 every 10 minutes). Of course, this doesn't come for free; we'd have to track additional state (jobs) on a per-broadcaster basis.

[1] Most likely this would consist of allocating a fresh `senderNonce` set for each "new" `recipientRand`, which is more expensive than putting a `senderNonce` into an existing set.

> O cannot simply maintain a set of old recipientRand values for winning tickets per stream because B1 could use a old recipientRand value for a winning ticket sent by B2

If the `recipientRand` construction includes the broadcaster address, then `recipientRand` can't practically be shared among broadcasters. For example,
```
recipientRand = HMAC(randSecret, randSeed | expiry | broadcasterAddress | ... )
```
With this construction, any submission from B1 would fail because `recipientRand` is used with B2's address. (See below on the `expiry` parameter.)

However, the larger point about thread-safe access still holds regardless of how `recipientRand` is stored (whether per-broadcaster or globally). To make things performant, we'll probably still need a few layers of locks, eg:

```golang
func AddToSet(recipientRand, senderNonce) {
  // technically nonce map checks should be wrapped by a read-lock
  // wrap with a write-lock if adding a new recipientRand to the nonce map
  // would be nice to have a concurrent map implementation
  senderNonceMutex.ReadLock()
  recipientRandNonces, exists = senderNonces[recipientRand]
  senderNonceMutex.Unlock()
  if !exists {
    recipientRandNonces = NewSet(..)
    senderNonceMutex.WriteLock()
    senderNonces[recipientRand] = recipientRandNonces
    senderNonceMutex.Unlock()
  }
  // Write-lock recipientRandNonces separately for concurrency at the `recipientRand` level
  // hopefully we can use a Set implementation with this logic already built in
  // maybe https://godoc.org/github.com/deckarep/golang-set
  recipientRandNonces.Mutex.WriteLock()
  exists = recipientRandNonces.Contains(senderNonce)
  if !exists {
    recipientRandNonces.Add(senderNonce)
  } // else reused nonce!
  recipientRandNonces.Mutex.Unlock()
}
```
This should give us reasonably fine-grained concurrency with the only global blocking happening when a `recipientRand` is added to the set. We can go even further by storing separate `senderNonces` maps per-broadcaster (was this the thinking behind making PM non-shareable between broadcasters?), but that step doesn't seem to be really necessary right now.

>  O might choose to refresh randSecret periodically in order to reset the recipientRand set.

We can also incorporate an expiration time in the `recipientRand` construction (see above). This allows O to flush state associated with old `recipientRands` after that expiry, such as sender nonces. This mitigates the need for a full flush if `randSecret` is refreshed: ongoing streams are not impacted, and the state space can be cleaned gradually rather than all-at-once.

Additionally, if winning tickets are [stored in the DB](https://github.com/livepeer/go-livepeer/issues/621), then SQLite gives us threadsafe concurrent access for free, without further in-memory bookkeeping of winning tickets. As long as `recipientRand` is indexed, the winning-ticket lookup should be quick.

**Comment by eladmallel** [comment]
> If the recipientRand construction includes the broadcaster address, then recipientRand can't practically be shared among broadcasters

This makes sense to me, so just trying to clarify assuming we add `broadcasterAddress` into the HMAC mix, why would we still need the global `senderNonces` map?

The reason that comes to me is the prevent the same B from replaying losing tickets across streams. Is that the reason you have in mind?

If so, perhaps we should have a map per B, which should further mitigate the impact inside of O that different Bs have on each other's streams.

> the concurrency limit effectively throttles the growth of the in-memory state space associated with a given B (100 every 10 minutes)

This makes a lot of sense to me, and I think this is something we should do. I'd add it under SS rather than MVS.

wdyt?

**Comment by j0sh** [comment]
> The reason that comes to me is the prevent the same B from replaying losing tickets across streams.

Yep, that's why we still need to track `senderNonces`.

> If so, perhaps we should have a map per B, which should further mitigate the impact inside of O that different Bs have on each other's streams.

We can do that, but it's another layer of state : `global map  of active Bs -> recipientRand map -> senderNonce set` . As mentioned earlier, this may still have other benefits such as finer grained concurrency and the throttling of "new" `recipientRand` values on a per-broadcaster basis to mitigate the memory consumption attack. I don't see it as absolutely necessary right now (perhaps for SS) but feel free to implement if that's what makes the most sense to you.

**Comment by yondonfu** [comment]
> We can also incorporate an expiration time in the recipientRand construction

Is `expiry` generated on a per `randSeed` basis? If so, I don't think this obviates the need to occasionally flush state by refreshing `randSecret` since O can only flush state on a per `randSeed` basis if it actually receives a ticket with `expiry` included as metadata (i.e. O might just be left hanging without receiving additional tickets). I do like the ability to flush state on a per `randSeed` basis though.

**Comment by j0sh** [comment]
> Is expiry generated on a per randSeed basis? If so, I don't think this obviates the need to occasionally flush state by refreshing randSecret since O can only flush state on a per randSeed basis

Yes, that was the idea. Agreed that flushing `randSecret` may still be useful on its own, but what other state depends on `randSecret` outside of `recipientRand`?

**Comment by eladmallel** [comment]
**This is an attempt to summarize our latest understanding of how we'd like to solve the random numbers coordination problem.** Once we get to a point of consensus I'll replace the original ticket description with the updated one (while preserving a record of the old for reference).

*Big thanks to @j0sh for helping us with valuable feedback!*

## Is your feature request related to a problem? Please describe.


__The Basics:__

* PM is based on B sending O payment tickets that are only redeemable under a certain probability (the probability is subject to B & O's negotiation).
* The function that determines if a ticket is redeemable (i.e. a winning ticket) takes two random numbers as inputs: one from B and one from O, named `senderNonce` and `recipientRand` respectively.

__Problem/Solution Flow:__

**Problem:** An obvious attack vector to defend against is B sending `senderNonce`s such that no ticket is redeemable. 
**Solution:** provide B with O's commitment to a `recipientRand` without revealing it to B, thus only O can determine for any `senderNonce` whether a given ticket is redeemable. We name this commitment `recipientRandHash`.

**Problem:** Any B can ask any O for their `OrchestratorInfo`, which should contain O's expected payment parameters, as well as a new `recipientRandHash`, so B can start streaming right away if B chooses to work with O. This means that O must be able to listen for tickets from Bs and determine if they are valid or not, and `recipientRandHash` is a key parameter to verify (for multiple reasons shared in this spec). How does O verify `recipientRandHash` is valid?
**Solution:** [HMAC](https://en.wikipedia.org/wiki/HMAC) to the rescue!
* Generate a `randSecret` every time O boots (and never share it).
* Generate a `randSeed` every time we want to generate a new `recipientRand`.
* Generate `recipientRand`: `recipientRand = HMAC(randSecret, randSeed)`
* `recipientRandHash = H(recipientRand)`
* Share both `recipientRandHash` and `randSeed` with B.

This approach allows O to confirm that a given ticket contains a valid `recipientRand` using HMAC:
* `recipientRand = HMAC(randSecret, ticket.randSeed)`
* `H(recipientRand) == ticket.recipientRandHash` should be true.

**Problem:** A new problem arises when O submits winning tickets to the `TicketBroker` smart contract: O must reveal `recipientRand` for the contract to verify the ticket is a winner before it settles the payment, thus making `recipientRand` visible to B as well (actually to all Bs).
**Solution:** Consists of two parts:
1. Making sure B1 cannot replay any tickets from B2
    1. Done by including B's address in the HMAC function, like so: `recipientRand = HMAC(randSecret, randSeed | b.address | ...)`
2. Making sure B1 cannot replay old `recipientRandHash` values to which they can attach a certainly losing `senderNonce`.
    1. Upon ticket redemption O generates a new `randSeed` and transmits to B a new pair of `randSeed` and `recipientRandHash` for B to include in new tickets. 
    2. O stores a list of all redeemed `recipientRand` values and any ticket that uses already-redeem values will be considered invalid.

**Problem:** The next problem arises from the asynchronous nature of B's and O's operations, e.g. B might be in the middle of sending a ticket with the old `recipientRand` commitment just as O is broadcasting the new `recipientRand`. 
**Solution:** We think O should provide some grace, allowing B to use the old value for some `N` number of tickets before, to provide a continuous transcoding service for all honest Bs.

**Problem:** The next problem is that given say a grace of `N=3` invalid tickets in a row, B could exploit O by sending an invalid ticket every other ticket, which would conform to the `N=3` threshold, but would result in 50% of tickets being invalid, which is obviously unwelcome. 
**Solution:** O should maintain a global ratio of `# invalid tickets / # all tickets` and a configurable threshold for this ratio. By default O will block any B that crosses this threshold.

*Note:* We agree that the absolute invalid ticket ratio is a more important and robust defense measure than the "in a row" approach. "In a row" is a nice-to-have addition that covers more cases.

**Problem:** If B knows O's threshold, B could game O by sending invalid tickets just below O's threshold. 
**Solution:** log (either to DB or a log file) all Bs' invalid ticket ratios to allow O's admins to monitor for such behaviors and potentially decide to block specific Bs. **TODO**: like to the relevant issue here.

**Problem:** Finally B could attempt to replay `senderNonce` values it believes did not yield a winning ticket for O (for the same `recipientRand` commitment).
**Solution:** 
* O will protect against this attack by maintaining a global set of `recipientRand` and `senderNonce` pairs, and block any ticket that tries to reuse an already-existing pair.
* This cache can be reset every time `randSecret` is reset.

**Problem:** We expect production-grade Os to be long-running processes, which would grow a large replay protection data structure in-memory and risk memory-based DDoS attacks and even max memory utilization after long honest interactions. How does O not run out of memory?
**Solution:**
* We are not solving this problem for MVS, and would like to dedicate more time to researching solutions.
* Bloom filter is a potential space-efficient solution we'd like to explore.
* We think it's a good idea to cap the number of concurrent streams any B can have open with O (e.g. 100)
* If no space-efficient solution is chosen, we will look more deeply into resetting `randSecret` at some frequency while O keeps running. We'd have to balance the benefit of memory cleanup against the cost of each `randSecret` reset resulting in some revenue loss due to the race condition between B's processes of receiving O's new rands and B's async sending of tickets.
* We could add an `expiry` param to the HMAC function, and upon receiving expired tickets delete those tickets' replay protection from memory (since the expiry becomes the new replay protection once the expiration time has passed).

**Future concern:**
* If and when we decouple tickets from segments (i.e. allowing a ticket to pay for multiple segments), we will also have to protect against B not sending enough tickets on time (e.g. with some timeout).

## Describe the solution you'd like

* O generates a new `randSecret` every time it boots.
* Any B wishing to work with O queries O for an `OrchestratorInfo` message, which contains new `randSeed` and `recipientRandHash` values derived from `randSecret` using HMAC:
    * `recipientRand = HMAC(randSecret, randSeed | b.address)`
    * `oInfo.recipientRandHash = H(recipientRand)`
    * `oInfo.randSeed = randSeed`
* Upon receiving a winning ticket O persists that ticket to DB (https://github.com/livepeer/go-livepeer/issues/621), to make sure it will be redeemed even if O crashes before the ticket is confirmed on-chain.
* When O decides to submit a winning ticket on-chain, O generates a new `randSeed` and `recipientRandHash` and sends them to the relevant B.
* O maintains a set of all redeem tickets' `recipientRand`s and checks every incoming ticket's `recipientRand` is not part of the set.
* O maintains a set of `recipientRand` and `senderNonce` pairs and checks every incoming ticket against it to prevert replays.
* O allows B to send `invalidTicketsInARowAllowed` invalid tickets in a row before disconnecting, to allow B time to receive and use the new `recipientRand` value (default setting `invalidTicketsInARowAllowed=3`).

### After MVS

* O maintains (on disk) an ongoing invalid tickets ratio per B as well as a `invalidTicketRatioThreshold` above which O rejects further video/tickets from B (`invalidTicketRatioThreshold` should be at least `invalidTicketsInARowAllowed/ticketWinningProbability`)
* All the memory optimizations and cleanups discussed above (Bloom filter, ticket expiry, routine cleanups, etc.)

## Describe alternatives you've considered

* We considered enforcing ticket order, but since currently tickets and segments are coupled, enforcing such order would hurt video performance, and since video performance is top priority, this solution seemed unfavorable.
* For other considerations best to follow the conversation thread on this ticket!

@j0sh @yondonfu does this summarize our discussions well? - please share any further feedback.

Thanks!

**Comment by j0sh** [comment]
@eladmallel This looks great, thanks for writing it up! Just a few notes:

> "In a row" is a nice-to-have addition that covers more cases.

Just curious, what cases are those?

At some point, we probably also should distinguish between tickets that are outright invalid (the segment shouldn't be processed at all; no grace period), and tickets that were valid but have since expired or had a winning `recipientRand` (grace period where O processes the segment anyway).

In general, using a ratio for invalid tickets feels a bit finicky in that each solution seems to present even more problems.

In lieu of a ratio, an absolute counter can be kept, and the counter reset under a few possible conditions:
* A ticket wins, necessitating a new `recipientRand`
* Time elapsed since the last failing ticket
* All jobs for a given B have timed out *or*
* All outstanding `recipientRand` values for a B have expired

This would also mitigate the need for any persistent on-disk storage. However, I still like the idea of some on-disk statistics simply for the analytic/forensic value.

> How does O not run out of memory?

This is probably implicitly stated, but just wanted to note that these solutions aren't mutually exclusive, and indeed would probably work better in combination with one another. For example:

* We only need one bloom filter if we combine the `senderNonce` with `recipientRand`, eg `key=H(recipientRand|senderNonce)` . However, we'd have to reset the bloom filter eventually -- which involves flushing all the state by resetting `randSecret`. This works, but it's a blunt approach.
* To mitigate this, can have separate `senderNonce` bloom filters for each `recipientRand` . These can expire piecemeal, with overall state-space growth managed by the concurrency cap.

**Comment by yondonfu** [comment]
> At some point, we probably also should distinguish between tickets that are outright invalid (the segment shouldn't be processed at all; no grace period), and tickets that were valid but have since expired or had a winning recipientRand (grace period where O processes the segment anyway).

Perhaps segments should be processed within a grace period for:

- Invalid `faceValue` and/or `winProb` since these values might have been acceptable at one point, but perhaps O now no longer accepts these values and B does not know about the new values yet
- Invalid `recipientRand` + `randSeed` since B might not have received the new pair after sending out a winning ticket yet due to async messaging

and segments should not be processed for:

- Invalid `senderNonce` i.e. a replay
- Other invalid fields in the ticket. Ex. invalid recipient address (i.e. not O's address), invalid timestamp (or whatever time indicator is being used), invalid signature from B, etc.

The general criteria for whether or not a grace period should be allowed is: could the invalid data in the ticket be a result of information that B has not yet received?

> In lieu of a ratio, an absolute counter can be kept, and the counter reset under a few possible conditions

👍 

> Time elapsed since the last failing ticket

How about valid tickets since the last acceptable invalid (assuming we make a distinction between acceptable invalid vs. unacceptable invalid) ticket?

**Comment by eladmallel** [comment]
> Just curious, what cases are those?

Gave it some more thought just now, and actually did not come up with any additional cases that a overall ratio would not sufficiently defend against. So I'm not advocating for this approach any longer.
 
**@yondonfu do you have any additional thoughts on this matter?**

> At some point, we probably also should distinguish between tickets that are outright invalid (the segment shouldn't be processed at all; no grace period)

> The general criteria for whether or not a grace period should be allowed is: could the invalid data in the ticket be a result of information that B has not yet received?

💯. This has always been the intention in my mind at least, but great to make it explicit that we only want to allow grace in those cases when it's expected that B needs some time to receive and use new params.

> Time elapsed since the last failing ticket

> How about valid tickets since the last acceptable invalid (assuming we make a distinction between acceptable invalid vs. unacceptable invalid) ticket?

I like the idea of a reset after a winning ticket or any timeout condition with B. What benefit do we get from the additional reset conditions above?

> This would also mitigate the need for any persistent on-disk storage. However, I still like the idea of some on-disk statistics simply for the analytic/forensic value.

Let's discuss a specific implementation here? - I was thinking a table with the following columns:

- B address
- Total tickets 
- Total winning tickets 
- Total outdated-rand-params tickets
- Total outdated-pricing-params tickets
- Total invalid-without-grace tickets

The moment we decide to reset the invalid tickets counter would be the same moment we decide to persist those counters to disk (async of course).

**wdyt?**

> This is probably implicitly stated, but just wanted to note that these solutions aren't mutually exclusive, and indeed would probably work better in combination with one another.

💯

**Comment by yondonfu** [comment]
Closing this because:

- A grace period is provided until a set of ticket params advertised by an orchestrator expires - a broadcaster can continue to use older ticket params until an expiration block in the future specified by the orchestrator
- Memory growth for storing revealed recipientRand values should no longer be a problem because we are no longer storing revealed recipientRand values. Instead, an orchestrator only reveals a recipientRand value on-chain after the ticket params that include the corresponding recipientRandHash expire
- Memory growth for storing used senderNonce values should no longer be a problem because we clear senderNonce value for a set of ticket params after the ticket params expire
